README
BinaryTree version 0.01
======================

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  blah blah blah

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2015 by Lin Yo-an

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.18.2 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension BTree.

0.02  Tue Apr  7 21:29:29 2015
    - Add traverse_* methods

0.01  Sat Apr  4 01:01:10 2015
	- original version; created by h2xs 1.23 with options



BTree/Makefile.PL
use 5.018002;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'BinaryTree::XS',
    VERSION_FROM      => 'lib/BinaryTree/XS.pm',
    PREREQ_PM         => {}, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/BinaryTree/XS.pm',
       AUTHOR         => 'Lin Yo-an <c9s@local>') : ()),
    LIBS              => [''], # e.g., '-lm'

    # TODO: Convert a perl hashref to the '-D" arguments?
    DEFINE            => '-DENABLE_DEBUG', # e.g., '-DHAVE_SOMETHING'
    INC               => '-I.', # e.g., '-I. -I/usr/include/other'
	# Un-comment this if you add C files to link with later:
    # OBJECT            => '$(O_FILES)', # link all the C files too
);


lib/Tree/Binary/XS.pm
package Tree::Binary::XS;

use 5.018002;
use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use Tree::Binary ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.02';

require XSLoader;
XSLoader::load('Tree::Binary::XS', $VERSION);

# Preloaded methods go here.

1;
__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

Tree::Binary::XS - Perl extension for manipulating binary tree structure

=head1 SYNOPSIS

  use Tree::Binary::XS;
  my $tree = Tree::Binary::XS->new({ by_key => 'id' });

  $tree->insert({ foo => 'bar', id => 11 });

  $tree->insert([{ foo => 'bar', id => 11 }, ... ]);

  $ret->exists(10);
  $ret->exists({ id => 10, 'name' => 'Bob' });

  # Use specified key instead of the key from payload
  $tree->insert(10, { foo => 'bar' });

  # to insert multiple keys one time.
  @ret = $tree->insert_those([{ id => 10, 'name' => 'Bob' },  { id => 3, 'name' => 'John' }, { id => 2, 'name' => 'Hank' } ]);

  $tree->update(10, { foo => 'bar' })

  $n = $tree->search(10);

  $tree->exists(10);
  $tree->exists({ foo => 'bar' , id => 10 });

  $tree->inorder_traverse(sub { 
        my ($key, $node) = @_;
    });

  $tree->postorder_traverse(sub { 
        my ($key, $node) = @_;
    });

  $tree->preorder_traverse(sub { 
        my ($key, $node) = @_;
    });

=head1 DESCRIPTION

Please note this extension is not compatible with the L<Tree::Binary> package,
this module was redesigned and simplified the interface of manipulating tree
structure.

=head1 FUNCTIONS

=over 4

=item $tree = Tree::Binary::XS->new({  by_key => $field_name })

The C<new> method constructs the Tree object, you may specify the C<by_key> option to let L<Tree::Binary::XS> 
get the key from the inserted objects.

=item $tree->insert(hashref $object)

Once you've defined the C<by_key>, you can simply pass the object to the insert method, for example:

    $tree->insert({
        id => 11,
        name => 'John',
    });

And C<11> will be the key of the object.

=item $tree->insert(IV key, hashref $object)

If you want to specify another key to insert the object, you may pass the C<key> as the first argument of the C<insert> method.

=item @ret = $tree->insert_those([ $obj1, $obj2, ... ])

=item $tree->update(IV key, $new_object)

=item $tree->exists(IV key)

=item $tree->exists(hashref $object)

=item $tree->search(IV key)

=item $tree->search(hashref $object)


=back


=head2 Tree Traversal

=over 4

=item $tree->postorder_traverse(sub { my ($key, $node) = @_; })

=item $tree->preorder_traverse(sub { my ($key, $node) = @_; })

=item $tree->inorder_traverse(sub { my ($key, $node) = @_; })

=cut

=head1 EXPORT

None by default.

=head1 SEE ALSO

L<Tree::Binary>, L<Tree::Binary::Search>

=head1 AUTHOR

Lin Yo-an, E<lt>c9s@localE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2015 by Lin Yo-an

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.18.2 or,
at your option, any later version of Perl 5 you may have available.


=cut


BTree/lib/BTree.pm
package BTree;

use 5.018002;
use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use BTree ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.01';

require XSLoader;
XSLoader::load('BTree', $VERSION);

# Preloaded methods go here.

1;
__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

BTree - Perl extension for blah blah blah

=head1 SYNOPSIS

  use BTree;
  my $tree = BTree->new({ by_key => 'id' });

  $tree->insert({ foo => 'bar', id => 11 });

  # to insert multiple keys one time.
  $tree->insert([{ foo => 'bar', id => 11 }, ... ]);

  $ret->exists(10);
  $ret->exists({ id => 10, 'name' => 'Bob' });

  # Use specified key instead of the key from payload
  $tree->insert(10, { foo => 'bar' });

  # Bulk insert
  @ret = $tree->insert_those([{ id => 10, 'name' => 'Bob' },  { id => 3, 'name' => 'John' }, { id => 2, 'name' => 'Hank' } ]);

  $tree->update(10, { foo => 'bar' })

  $n = $tree->search(10);

  $tree->exists(10);
  $tree->exists({ foo => 'bar' , id => 10 });


=head1 DESCRIPTION

Stub documentation for BTree, created by h2xs. It looks like the
author of the extension was negligent enough to leave the stub
unedited.

Blah blah blah.

=head2 EXPORT

None by default.



=head1 SEE ALSO

Mention other useful documentation such as the documentation of
related modules or operating system documentation (such as man pages
in UNIX), or any relevant external documentation such as RFCs or
standards.

If you have a mailing list set up for your module, mention it here.

If you have a web site set up for your module, mention it here.

=head1 AUTHOR

Lin Yo-an, E<lt>c9s@localE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2015 by Lin Yo-an

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.18.2 or,
at your option, any later version of Perl 5 you may have available.


=cut
